contrib/generic_proxy/filters/network/source/codecs/http1/config.h

```diff
@@ -0,0 +1,392 @@
+    header_parsing_state_ = HeaderParsingState::Done;
+    return onHeadersCompleteImpl();
+  }
+  void bufferBody(const char* data, size_t length) override { buffered_body_.add(data, length); }
+  Http::Http1::CallbackResult onMessageComplete() override { return onMessageCompleteImpl(); }
+  void onChunkHeader(bool is_final_chunk) override {
+    if (is_final_chunk) {
+      dispatchBufferedBody(false);
+    }
+  }
+
+  virtual Http::Http1::CallbackResult onMessageBeginImpl() PURE;
+  virtual void onUrlImpl(const char* data, size_t length) PURE;
+  virtual void onStatusImpl(const char* data, size_t length) PURE;
+  virtual Http::Http1::CallbackResult onHeadersCompleteImpl() PURE;
+  virtual Http::Http1::CallbackResult onMessageCompleteImpl() PURE;
+
+  void completeCurrentHeader() {
+    current_header_value_.rtrim();
+    current_header_field_.inlineTransform([](char c) { return absl::ascii_tolower(c); });
+    headerMap().addViaMove(std::move(current_header_field_), std::move(current_header_value_));
+
+    ASSERT(current_header_field_.empty());
+    ASSERT(current_header_value_.empty());
+  }
+
+  bool decodeBuffer(Buffer::Instance& buffer);
+
+  void dispatchBufferedBody(bool end_stream);
+  bool bufferedBodyOverflow();
+
+  virtual Http::HeaderMap& headerMap() PURE;
+
+  virtual void onDecodingSuccess(StreamFramePtr&& frame) PURE;
+  virtual void onDecodingFailure() PURE;
+
+protected:
+  enum class HeaderParsingState { Field, Value, Done };
+
+  Envoy::Buffer::OwnedImpl decoding_buffer_;
+  Envoy::Buffer::OwnedImpl encoding_buffer_;
+
+  Buffer::OwnedImpl buffered_body_;
+
+  Http::Http1::ParserPtr parser_;
+  Http::HeaderString current_header_field_;
+  Http::HeaderString current_header_value_;
+  HeaderParsingState header_parsing_state_{HeaderParsingState::Field};
+
+  const bool single_frame_mode_{};
+  const uint32_t max_buffer_size_{};
+
+  bool deferred_end_stream_headers_{};
+};
+
+class Http1ServerCodec : public Http1CodecBase, public ServerCodec {
+public:
+  Http1ServerCodec(bool single_frame_mode, uint32_t max_buffer_size)
+      : Http1CodecBase(single_frame_mode, max_buffer_size, true) {}
+
+  Http::Http1::CallbackResult onMessageBeginImpl() override;
+  void onUrlImpl(const char* data, size_t length) override {
+    ASSERT(active_request_.has_value());
+    ASSERT(active_request_->request_headers_ != nullptr);
+    active_request_->request_headers_->setPath(absl::string_view(data, length));
+  }
+  void onStatusImpl(const char*, size_t) override {}
+  Http::Http1::CallbackResult onHeadersCompleteImpl() override;
+  Http::Http1::CallbackResult onMessageCompleteImpl() override;
+
+  Http::HeaderMap& headerMap() override { return *active_request_->request_headers_; }
+
+  void setCodecCallbacks(ServerCodecCallbacks& callbacks) override { callbacks_ = &callbacks; }
+  void decode(Envoy::Buffer::Instance& buffer, bool) override {
+    if (!decodeBuffer(buffer)) {
+      callbacks_->onDecodingFailure();
```

Comment:
It feels like we just reset the connection for any decode error or incorrect request. It won't return any normal non-200 message to tell the client what happened?

Reply:
May be we can send some local response to tell the error in the future. If you agree, I will do it do it in a separated PR.

Reply:
yea, I'm ok with it now, since I know this is probably ok for your usecase. I think we just need to document them, let other people knowing how http1 codec behavior is.

Reply:
New comments were added. And I also added a TODO. Will create a new PR in next one or two weeks. :)